core/audits/preload-lcp-image.js

```diff
@@ -116,6 +110,29 @@ class PreloadLCPImageAudit extends Audit {
     };
   }
 
+  /**
+   * Match the LCP event with the paint event to get the URL of the image actually painted.
+   * This could differ from the `ImageElement` associated with the nodeId if e.g. the LCP
+   * was a pseudo-element associated with a node containing a smaller background-image.
+   * @param {LH.Trace} trace
+   * @param {LH.Artifacts.ProcessedNavigation} processedNavigation
+   * @return {string | undefined}
+   */
+  static getLcpUrl(trace, processedNavigation) {
+    const lcpEvent = processedNavigation.largestContentfulPaintAllFramesEvt;
+    if (!lcpEvent) return;
+
+    const lcpImagePaintEvent = trace.traceEvents.filter(e => {
+      return e.name === 'LargestImagePaint::Candidate' &&
+          e.args.frame === lcpEvent.args.frame &&
+          e.args.data?.DOMNodeId === lcpEvent.args.data?.nodeId &&
+          e.args.data?.size === lcpEvent.args.data?.size;
+    // Get last candidate, in case there was more than one.
+    }).sort((a, b) => b.ts - a.ts)[0];
```

Comment:
wouldn't this just be the last item in this array? assuming trace events are already sorted by ts. if so, could also do `findLast` as a micro opt (as well as being more self-documenting)

Reply:
this is going back to the raw trace, so they will be mostly sorted, but not completely (unlike e.g. `processedTrace.processEvents` [which we sort ourselves](https://github.com/GoogleChrome/lighthouse/blob/421af3866d51229f78f826b2ed4702aa107cbaa9/core/lib/tracehouse/trace-processor.js#L654-L655))